<a href="https://colab.research.google.com/github/songwoojin04/machine2/blob/main/3week/weather_regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [65]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import LabelEncoder

weather = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/3주차/weather.csv")
weather

Date  Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0     11/1/2007  Canberra      8.0     24.3       0.0          3.4       6.3   
1     11/2/2007  Canberra     14.0     26.9       3.6          4.4       9.7   
2     11/3/2007  Canberra     13.7     23.4       3.6          5.8       3.3   
3     11/4/2007  Canberra     13.3     15.5      39.8          7.2       9.1   
4     11/5/2007  Canberra      7.6     16.1       2.8          5.6      10.6   
..          ...       ...      ...      ...       ...          ...       ...   
361  10/27/2008  Canberra      9.0     30.7       0.0          7.6      12.1   
362  10/28/2008  Canberra      7.1     28.4       0.0         11.6      12.7   
363  10/29/2008  Canberra     12.5     19.9       0.0          8.4       5.3   
364  10/30/2008  Canberra     12.5     26.9       0.0          5.0       7.1   
365  10/31/2008  Canberra     12.3     30.2       0.0          6.0      12.6   

    WindGustDir  WindGustSpeed WindDir9am  ... Humidity3pm  Pressure9am  \
0            NW           30.0         SW  ...          29       1019.7   
1           ENE           39.0          E  ...          36       1012.4   
2            NW           85.0          N  ...          69       1009.5   
3            NW           54.0        WNW  ...          56       1005.5   
4           SSE           50.0        SSE  ...          49       1018.3   
..          ...            ...        ...  ...         ...          ...   
361         NNW           76.0        SSE  ...          15       1016.1   
362           N           48.0        NNW  ...          22       1020.0   
363         ESE           43.0        ENE  ...          47       1024.0   
364          NW           46.0        SSW  ...          39       1021.0   
365          NW           78.0         NW  ...          13       1009.6   

     Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday  RISK_MM  \
0         1015.0         7         7     14.4     23.6         No      3.6   
1         1008.4         5         3     17.5     25.7        Yes      3.6   
2         1007.2         8         7     15.4     20.2        Yes     39.8   
3         1007.0         2         7     13.5     14.1        Yes      2.8   
4         1018.5         7         7     11.1     15.4        Yes      0.0   
..           ...       ...       ...      ...      ...        ...      ...   
361       1010.8         1         3     20.4     30.0         No      0.0   
362       1016.9         0         1     17.2     28.2         No      0.0   
363       1022.8         3         2     14.5     18.3         No      0.0   
364       1016.2         6         7     15.8     25.9         No      0.0   
365       1009.2         1         1     23.8     28.6         No      0.0   

     RainTomorrow  
0             Yes  
1             Yes  
2             Yes  
3             Yes  
4              No  
..            ...  
361            No  
362            No  
363            No  
364            No  
365            No  

[366 rows x 24 columns]

In [66]:
weather.isnull().sum()

Date              0
Location          0
MinTemp           0
MaxTemp           0
Rainfall          0
Evaporation       0
Sunshine          3
WindGustDir       3
WindGustSpeed     2
WindDir9am       31
WindDir3pm        1
WindSpeed9am      7
WindSpeed3pm      0
Humidity9am       0
Humidity3pm       0
Pressure9am       0
Pressure3pm       0
Cloud9am          0
Cloud3pm          0
Temp9am           0
Temp3pm           0
RainToday         0
RISK_MM           0
RainTomorrow      0
dtype: int64

In [67]:
weather = weather.dropna()
weather.isnull().sum()

Date             0
Location         0
MinTemp          0
MaxTemp          0
Rainfall         0
Evaporation      0
Sunshine         0
WindGustDir      0
WindGustSpeed    0
WindDir9am       0
WindDir3pm       0
WindSpeed9am     0
WindSpeed3pm     0
Humidity9am      0
Humidity3pm      0
Pressure9am      0
Pressure3pm      0
Cloud9am         0
Cloud3pm         0
Temp9am          0
Temp3pm          0
RainToday        0
RISK_MM          0
RainTomorrow     0
dtype: int64

In [68]:
weather.columns

Index(['Date', 'Location', 'MinTemp', 'MaxTemp', 'Rainfall', 'Evaporation',
       'Sunshine', 'WindGustDir', 'WindGustSpeed', 'WindDir9am', 'WindDir3pm',
       'WindSpeed9am', 'WindSpeed3pm', 'Humidity9am', 'Humidity3pm',
       'Pressure9am', 'Pressure3pm', 'Cloud9am', 'Cloud3pm', 'Temp9am',
       'Temp3pm', 'RainToday', 'RISK_MM', 'RainTomorrow'],
      dtype='object')

In [69]:
weather = weather.drop("RainTomorrow", axis=1)

In [70]:
label_encoders = {}
for column in weather.columns:
    if weather[column].dtype == 'object':
        le = LabelEncoder()
        weather[column] = le.fit_transform(weather[column])
        label_encoders[column] = le

In [71]:
weather.columns

Index(['Date', 'Location', 'MinTemp', 'MaxTemp', 'Rainfall', 'Evaporation',
       'Sunshine', 'WindGustDir', 'WindGustSpeed', 'WindDir9am', 'WindDir3pm',
       'WindSpeed9am', 'WindSpeed3pm', 'Humidity9am', 'Humidity3pm',
       'Pressure9am', 'Pressure3pm', 'Cloud9am', 'Cloud3pm', 'Temp9am',
       'Temp3pm', 'RainToday', 'RISK_MM'],
      dtype='object')

In [72]:
X = weather.drop('Rainfall',axis=1)
X.head()

Date  Location  MinTemp  MaxTemp  Evaporation  Sunshine  WindGustDir  \
0    59         0      8.0     24.3          3.4       6.3            7   
1    70         0     14.0     26.9          4.4       9.7            1   
2    81         0     13.7     23.4          5.8       3.3            7   
3    83         0     13.3     15.5          7.2       9.1            7   
4    84         0      7.6     16.1          5.6      10.6           10   

   WindGustSpeed  WindDir9am  WindDir3pm  ...  Humidity9am  Humidity3pm  \
0           30.0          12           7  ...           68           29   
1           39.0           0          13  ...           80           36   
2           85.0           3           5  ...           82           69   
3           54.0          14          13  ...           62           56   
4           50.0          10           2  ...           68           49   

   Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday  \
0       1019.7       1015.0         7         7     14.4     23.6          0   
1       1012.4       1008.4         5         3     17.5     25.7          1   
2       1009.5       1007.2         8         7     15.4     20.2          1   
3       1005.5       1007.0         2         7     13.5     14.1          1   
4       1018.3       1018.5         7         7     11.1     15.4          1   

   RISK_MM  
0      3.6  
1      3.6  
2     39.8  
3      2.8  
4      0.0  

[5 rows x 22 columns]

In [73]:
y = weather['Rainfall']
y.head()

0     0.0
1     3.6
2     3.6
3    39.8
4     2.8
Name: Rainfall, dtype: float64

In [74]:
X_train, X_test, y_train, y_test=train_test_split(X, y, test_size=0.2, shuffle=True,random_state=12)

print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

(262, 22) (262,)
(66, 22) (66,)


In [75]:
model = LinearRegression()
model.fit(X_train, y_train)
ly_preds = model.predict(X_test)

print(mean_squared_error(y_test, ly_preds))

22.77359387258412


해당 수의 의미는 테스트 데이터에서 예측값과 실제값의 차이를 제곱해서 평균 낸 값을 의미하며 값이 0에 가까울수록 모델이 잘 맞춘다는 의미입니다.

즉 예측값과 실제값 사이에 평균적으로 제곱 오차가 22.77